remote.c

```diff
@@ -2163,6 +2163,150 @@ do_del_tsig(RES* ssl, xfrd_state_type* xfrd, char* arg) {
+		(void)ssl_printf(ssl, "unable to open cookie secret file %s: %s",
+		                 secret_file, strerror(errno));
+		return 0;
+	}
+	for(size_t i = 0; i < nsd->cookie_count; i++) {
+		struct cookie_secret const* cs = &nsd->cookie_secrets[i];
+		ssize_t const len = hex_ntop(cs->cookie_secret, NSD_COOKIE_SECRET_SIZE,
+																 secret_hex, sizeof(secret_hex));
+		(void)len; /* silence unused variable warning with -DNDEBUG */
+		assert( len == NSD_COOKIE_SECRET_SIZE * 2 );
+		secret_hex[NSD_COOKIE_SECRET_SIZE * 2] = '\0';
+		fprintf(f, "%s\n", secret_hex);
+	}
+	fclose(f);
+	return 1;
+}
+
+static void
+do_activate_cookie_secret(RES* ssl, xfrd_state_type* xrfd, char* arg) {
+	nsd_type* nsd = xrfd->nsd;
+	(void)arg;
+
+	if(nsd->cookie_count <= 1 ) {
+		(void)ssl_printf(ssl, "error: no staging cookie secret to activate\n");
+		return;
+	}
+	if(!nsd->options->cookie_secret_file || !nsd->options->cookie_secret_file[0]) {
+		(void)ssl_printf(ssl, "error: no cookie secret file configured\n");
+		return;
+	}
+	if(!cookie_secret_file_dump(ssl, nsd)) {
+		(void)ssl_printf(ssl, "error: writing to cookie secret file: \"%s\"\n",
+				nsd->options->cookie_secret_file);
+		return;
+	}
+	activate_cookie_secret(nsd);
+	(void)cookie_secret_file_dump(ssl, nsd);
+	task_new_activate_cookie_secret(xfrd->nsd->task[xfrd->nsd->mytask],
+	    xfrd->last_task);
+	xfrd_set_reload_now(xfrd);
+	send_ok(ssl);
+}
+
+static void
+do_drop_cookie_secret(RES* ssl, xfrd_state_type* xrfd, char* arg) {
+	nsd_type* nsd = xrfd->nsd;
+	(void)arg;
+
+	if(nsd->cookie_count <= 1 ) {
+		(void)ssl_printf(ssl, "error: can not drop the currently active cookie secret\n");
+		return;
+	}
+	if(!nsd->options->cookie_secret_file || !nsd->options->cookie_secret_file[0]) {
+		(void)ssl_printf(ssl, "error: no cookie secret file configured\n");
+		return;
+	}
+	if(!cookie_secret_file_dump(ssl, nsd)) {
+		(void)ssl_printf(ssl, "error: writing to cookie secret file: \"%s\"\n",
+				nsd->options->cookie_secret_file);
+		return;
+	}
+	drop_cookie_secret(nsd);
+	(void)cookie_secret_file_dump(ssl, nsd);
+	task_new_drop_cookie_secret(xfrd->nsd->task[xfrd->nsd->mytask],
+	    xfrd->last_task);
+	xfrd_set_reload_now(xfrd);
+	send_ok(ssl);
+}
+
+static void
+do_add_cookie_secret(RES* ssl, xfrd_state_type* xrfd, char* arg) {
+	nsd_type* nsd = xrfd->nsd;
+	uint8_t secret[NSD_COOKIE_SECRET_SIZE];
+
+	if(*arg == '\0') {
+		(void)ssl_printf(ssl, "error: missing argument (cookie_secret)\n");
+		return;
+	}
+	if(strlen(arg) != 32) {
+		(void)ssl_printf(ssl, "invalid cookie secret: invalid argument length\n");
+		(void)ssl_printf(ssl, "please provide a 128bit hex encoded secret\n");
+		memset(arg, 0xdd, strlen(arg));
+		return;
+	}
+	if(hex_pton(arg, secret, NSD_COOKIE_SECRET_SIZE) != NSD_COOKIE_SECRET_SIZE ) {
+		(void)ssl_printf(ssl, "invalid cookie secret: parse error\n");
+		(void)ssl_printf(ssl, "please provide a 128bit hex encoded secret\n");
+		memset(arg, 0xdd, strlen(arg));
```

Comment:
If you memset 'arg' you should also clear the 'secret' array, which contains partial results. Also for memset calls for arg further in this function, there should be a clear of the 'secret' array, as well, if you want to erase parts of that data.

Reply:
resolved in aa3bca3